### ISSUE TITLE
feature request - cheaper and mules

### ISSUE THREAD

**Issue by Lujeni** [unbit/uwsgi#1002]
Hi,

could uWSGI support a cheaper-like subsystem that I can use to dynamically spawn mules please ?
It's seems technically possible (the spooler already have something similar in 2.1).

Thanks for your works.

Lujeni

**Comment by ramnes** [comment]
:+1:

**Comment by btall** [comment]
:+1:

**Comment by ultrabug** [comment]
Yes Roberto please ! Tell us if we can help (tho we're no C++ devs), but you're sure to have some testers here :) :+1:

**Comment by unbit** [comment]
Ok, it is clear that France votes for cheaper mules :)

**Comment by ultrabug** [comment]
ahah yeah, hurrah ! :v: 

PS: we wanted to discuss this with you at EP2015 but ... :(

**Comment by unbit** [comment]
As considering a mule idle or busy is basically impossible (as they could have a brain) i think the only viable approach is exposing a mule_spawn() function to the api. Or i am missing something ? Obviously if you use mules only for signals and messages we could use it for automatic process management.

**Comment by ultrabug** [comment]
> As considering a mule idle or busy is basically impossible

I'm not sure what you mean, for us for example mules are processors which basically implement a sort of infinite loop to process jobs, so we definitively can have an idle and busy mule (aka waiting for a job or processing one)

> the only viable approach is exposing a mule_spawn() function to the api

This means that you would allow users to call a sort of uwsgi.mule_spawn() in their code and implement the logic of spawning (and killing) mules themselves ?

> Obviously if you use mules only for signals and messages we could use it for automatic process management.

Not sure I follow, must be because we don't use mule signals and stuff like that at all

**Comment by ultrabug** [comment]
My first and basic idea was to expose 2 functions to the users
* mule_idle()
* mule_busy()

And then allow a configuration in the INI with a threshold for saying something like
* mules_default = 5
* mules_max = 50
* mules_min = 1
* mules_spawn = 2
* mules_threshold = 10 (seconds)

And implement a logic where if all the currently number of spawned mules are busy for `mules_threshold` seconds, spawn `mules_spawn` new mules with a maximum of `mules_max` mules.

Does that make any sense ?
